lib/Sema/TypeCheckStmt.cpp

```diff
@@ -290,15 +290,167 @@ static void tryDiagnoseUnnecessaryCastOverOptionSet(ASTContext &Ctx,
+    bool isContinue, DeclContext *dc,
+    ArrayRef<LabeledStmt *> oldActiveLabeledStmts) {
+  TopCollection<unsigned, LabeledStmt *> labelCorrections(3);
+
+  // Retrieve the active set of labeled statements.
+  // FIXME: Once everything uses ASTScope lookup, \c oldActiveLabeledStmts
+  // can go away.
+  SmallVector<LabeledStmt *, 4> activeLabeledStmts;
+  if (ctx.LangOpts.EnableASTScopeLookup) {
+    activeLabeledStmts = ASTScope::lookupLabeledStmts(sourceFile, loc);
+  } else {
+    activeLabeledStmts.insert(
+        activeLabeledStmts.end(),
+        oldActiveLabeledStmts.rbegin(), oldActiveLabeledStmts.rend());
+  }
+
+  // Pick the nearest break target that matches the specified name.
+  if (targetName.empty()) {
+    for (auto labeledStmt : activeLabeledStmts) {
+      // 'break' with no label looks through non-loop structures
+      // except 'switch'.
+      // 'continue' ignores non-loop structures.
+      if (!labeledStmt->requiresLabelOnJump() &&
+          (!isContinue || labeledStmt->isPossibleContinueTarget())) {
+        return labeledStmt;
+      }
+    }
+  } else {
+    // Scan inside out until we find something with the right label.
+    for (auto labeledStmt : activeLabeledStmts) {
+      if (targetName == labeledStmt->getLabelInfo().Name) {
+        // Continue cannot be used to repeat switches, use fallthrough instead.
+        if (isContinue && !labeledStmt->isPossibleContinueTarget()) {
+          ctx.Diags.diagnose(
+              loc, diag::continue_not_in_this_stmt,
+              isa<SwitchStmt>(labeledStmt) ? "switch" : "if");
+          return nullptr;
+        }
+
+        return labeledStmt;
+      }
+
+      unsigned distance =
+        TypeChecker::getCallEditDistance(
+            DeclNameRef(targetName),
+            labeledStmt->getLabelInfo().Name,
+            TypeChecker::UnreasonableCallEditDistance);
+      if (distance < TypeChecker::UnreasonableCallEditDistance)
+        labelCorrections.insert(distance, std::move(labeledStmt));
+    }
+    labelCorrections.filterMaxScoreRange(
+      TypeChecker::MaxCallEditDistanceFromBestCandidate);
+  }
+
+  // If we're in a defer, produce a tailored diagnostic.
+  if (isDefer(dc)) {
+    ctx.Diags.diagnose(
+        loc, diag::jump_out_of_defer, isContinue ? "continue": "break");
+    return nullptr;
+  }
+
+  if (targetName.empty()) {
+    // If we're dealing with an unlabeled break inside of an 'if' or 'do'
+    // statement, produce a more specific error.
+    if (!isContinue &&
+        llvm::any_of(activeLabeledStmts,
+                     [&](Stmt *S) -> bool {
+                       return isa<IfStmt>(S) || isa<DoStmt>(S);
+                     })) {
+      ctx.Diags.diagnose(
+          loc, diag::unlabeled_break_outside_loop);
+      return nullptr;
+    }
+
+    // Otherwise produce a generic error.
+    ctx.Diags.diagnose(
+        loc,
+        isContinue ? diag::continue_outside_loop : diag::break_outside_loop);
+    return nullptr;
```

Comment:
This function is way too long. It does not fit onto a single screen on my 4K monitor. I harbor the strong intuition that it will be very difficult for many folks to understand and maintain, even though you have the horsepower to deal with it. I realize you have already merged this PR, but I hope you might go back and break it up.

Reply:
Simplified in https://github.com/apple/swift/pull/33278